### ISSUE TITLE
CPU & GPU usage

### ISSUE THREAD

**Issue by nicolasnoble** [saharmor/dalle-playground#17]
Apologies if this isn't the appropriate venue to ask these questions.

I've been toying a bit with this, but I'm seeing two problems so far:

 - The CPU seems heavily underused. Right now the process is only using a single core out of my 12-cores machine, and I couldn't figure out a way to indicate to use more resources.
  - The GPU is totally unused. In fact, there's a message at the beginning talking about it:
```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
--> Starting DALL-E Server. This might take up to two minutes.
```

Running with `TF_CPP_MIN_LOG_LEVEL=0` doesn't give much more info:
```
2022-06-06 13:07:53.761716: I external/org_tensorflow/tensorflow/core/util/util.cc:168] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2022-06-06 13:07:53.761954: I external/org_tensorflow/tensorflow/core/tpu/tpu_initializer_helper.cc:259] Libtpu path is: libtpu.so
2022-06-06 13:07:53.785866: I external/org_tensorflow/tensorflow/core/util/util.cc:168] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2022-06-06 13:07:54.419499: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2022-06-06 13:07:55.545835: I external/org_tensorflow/tensorflow/compiler/xla/service/service.cc:174] XLA service 0x555e2489cb80 initialized for platform Interpreter (this does not guarantee that XLA will be used). Devices:
2022-06-06 13:07:55.545853: I external/org_tensorflow/tensorflow/compiler/xla/service/service.cc:182]   StreamExecutor device (0): Interpreter, <undefined>
2022-06-06 13:07:55.547821: I external/org_tensorflow/tensorflow/compiler/xla/pjrt/tfrt_cpu_pjrt_client.cc:176] TfrtCpuClient created.
2022-06-06 13:07:55.548231: I external/org_tensorflow/tensorflow/stream_executor/tpu/tpu_platform_interface.cc:74] No TPU platform found.
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
--> Starting DALL-E Server. This might take up to two minutes.
```

I did follow the instructions to install pytorch, including the verification that cuda is detectable:
```
$ python
Python 3.9.12 (main, Apr  5 2022, 06:56:58) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.cuda.is_available()
True
```

To no avail.

Feel free to redirect me to other resources if this isn't the right medium to discuss the problem.

**Comment by mikaczma** [comment]
Try to build it using docker. Works like charm after all env vars are set properely.
I have tested it to use nvidia card using docker.

**Comment by saharmor** [comment]
+1 to @mikaczma's suggestion. Can you please try that and report back?

**Comment by nicolasnoble** [comment]
The work environment I was trying to run this is disallowing the usage of docker, unfortunately. Using docker at home is fine, sure.

**Comment by saharmor** [comment]
Did you try using the [Google Colab notebook](https://colab.research.google.com/github/saharmor/dalle-playground/blob/main/backend/dalle_playground_backend.ipynb)?

**Comment by nikisalli** [comment]
I was able to run dalle mega full on my laptop with 32GB ram by adding "jax.config.update('jax_platform_name', 'cpu')" after the imports in dalle_model.py

it takes a veeeery long time to generate just one image (2 hours or more on my i7 8750h) but it works

also it looks like it defaults to single thread for dalle mega full but goes correctly in multi thread mode for dalle mega and dalle mini

**Comment by fschwiet** [comment]
Hi @nicolasnoble sorry can I ask a newb question? How do I get  TF_CPP_MIN_LOG_LEVEL=0 to work? When I try to set it, I don't get extra output:

```
root@DESKTOP-OB9D3NI:~/code/dalle-playground/backend# set TF_CPP_MIN_LOG_LEVEL=0
root@DESKTOP-OB9D3NI:~/code/dalle-playground/backend# TF_CPP_MIN_LOG_LEVEL=0
root@DESKTOP-OB9D3NI:~/code/dalle-playground/backend# python3 app.py 8080
--> Starting DALL-E Server. This might take up to two minutes.
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
```

**Comment by joeltgray** [comment]
@fschwiet The easiest way to set it is by setting it from within the app.py file, at the top near the imports enter:
```
import os  
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '0'
```

Additionally if you want some more info you can turn debugging on for flask, on the last line of the app.py file change debug false to true:
`app.run(host="0.0.0.0", port=int(sys.argv[1]), debug=False) `
Change to:
`app.run(host="0.0.0.0", port=int(sys.argv[1]), debug=True)`

**Comment by fschwiet** [comment]
Thank you, will give it a try.

I wanted to know that though absl was reporting no GPU found that GPU usage does go up to 30-40% while I am generating images. My setup might be an outlier, running the app via Windows Subsystem for Linux on Windows 10. I see WSL doesn't support graphical UIs on Windows 10 but does on 11, so I'm upgrading now to see if that makes a difference. Maybe WLS is limiting the resources available to the app (CPU usage goes up to 50%).

**Comment by Ly-Zxzy** [comment]
I also had _WARNING:absl:No GPU/TPU found, falling back to CPU._ whilst using WSL on Windows 11 - I'm just amateurishly tinkering about so may have missed something, but I was able to resolve it by installing CUDA Toolkit separately (I used version 11.3 specifically, and for WSL the Ubuntu-WSL version not the normal Ubuntu one).

I believe the pytorch installation includes the toolkit, which means that pytorch is able to see it, but JAX wasn't seeing it until I installed the toolkit standalone - it errored when trying the below even though the torch test came back true:
Python: `import torch, jax; print(torch.cuda.is_available()); print(jax.devices())`

Once I installed the toolkit separately and ensured the cuda-11.* folder it places was added to %PATH% (https://docs.nvidia.com/cuda/archive/11.3.0/cuda-installation-guide-linux/index.html#post-installation-actions), it then seemed to pick up the GPU properly.

I then also needed cuDNN and to install its libraries to resolve a follow-up issue.
Once I did that, it seems to all be working and generates images within a few seconds.

**Comment by TechnicJelle** [comment]
I'm trying to do the same as you described, but it seems like it's not working for me. I have installed the CUDA Toolkit, although somehow it did install 11.7, instead of 11.3, but I don't _think_ that should be the problem?.
I've added the `cuda-11.7` thing to the path, as the Post-installation Actions say I have to do. I put it in my `.profile` file with `PATH=/usr/local/cuda-11.7/bin${PATH:+:${PATH}}`.
When I do 
```bash
$ echo $PATH
/usr/local/cuda-11.7/bin:/home/technicjelle/.local/bin: etc. etc.
```
you can see that it is there on the path.
But when I go into Python and run the one-liner you sent:
```python
$ python
Python 3.8.10 (default, Mar 15 2022, 12:22:08)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch, jax; print(torch.cuda.is_available()); print(jax.devices())
True
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
[CpuDevice(id=0)]
```

Also running on Ubuntu WSL, but on Windows 10, because apparently that also supports CUDA in WSL now:
```bash
$ nvidia-smi -L
GPU 0: NVIDIA GeForce RTX 3070 Laptop GPU (UUID: GPU-a806ee65-1125-2137-1947-a968a545aa27)
```

**Comment by fschwiet** [comment]
@Ly-Zxzy What CPU and GPU usage do you reach when generating images?

**Comment by Ly-Zxzy** [comment]
Hmm, might also be worth also upgrading jax then if you haven't yet.
`pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html`

That's the only thing I can recall potentially doing outside of the CUDA toolkit instructions that would have otherwise also fixed it. Perhaps also try shutting down and relaunching WSL (and ensure to re-add cuda-11.*/bin to path after if it doesn't stick).

If not, might be worth trying toolkit 11.3 regardless just in case there is some weirdness going on with the versioning, I picked 11.3 at first since it's the one pytorch uses. Otherwise, I'm unfortunately not too sure.
https://developer.nvidia.com/cuda-toolkit-archive

**Comment by Ly-Zxzy** [comment]
> @Ly-Zxzy What CPU and GPU usage do you reach when generating images?

Going by Windows Task Manager, 95%ish on my 3090, 20-25%ish on my i9-12900K.

**Comment by TechnicJelle** [comment]
> Hmm, might also be worth also upgrading jax then if you haven't yet. `pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html`

THAT DID IT!
```python
$ python
Python 3.8.10 (default, Mar 15 2022, 12:22:08)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch, jax; print(torch.cuda.is_available()); print(jax.devices())
True
[GpuDevice(id=0, process_index=0)]
```
Thanks a lot! I'm now going to try the actual dalle backend

**Comment by Ly-Zxzy** [comment]
> > Hmm, might also be worth also upgrading jax then if you haven't yet. `pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html`
> 
> THAT DID IT!

Ah glad to hear, I could have sworn I tried upgrading jax separately before and that on its own didn't work, so I assume it needs both that done _and_ the toolkit installed separately (or I just messed it up another way the first time around somehow, quite possible, not going to nuke WSL now to find out - but if someone else needs to then... ).

And yep, all I tried for the DNN library issue was installing cuDNN + runtime & developer libraries for that, as detailed in the install guide (https://docs.nvidia.com/deeplearning/cudnn/install-guide/index.html) so hopefully that should do the trick for it. I had issues apt-getting the libraries so had to download them manually from the source then install using the .deb file locally, in case you get the same thing.

**Comment by fschwiet** [comment]
I'm redoing my WSL install and had a question, how did you install pytorch? https://pytorch.org/get-started/locally/ tells me to run:

> pip3 install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu113

But I wonder if I should change it to 117 since thats the version of cuda I have.

**Comment by Ly-Zxzy** [comment]
Yes, I just used that standard provided install command, I don't think there's a version of pytorch specifically for 117 - they only incrementally update the cuda version for it from what I gathered.

If I'm assuming correctly, pytorch includes its own bundled cuda version which it'll just use by itself, and then you can install the cuda toolkit standalone for anything else to use (e.g. JAX).

If the cuda bundled with pytorch is supposed to be sufficient for everything, that would explain why we're running into these issues as it's evidently not finding it properly, but I wouldn't think that's the case (pretty sure it's not a full install)?

**Comment by Ly-Zxzy** [comment]
Scrolling up in the console, it appears I also received preliminary OOM errors after launching it, but after it finished starting the server it seemed to still all be working for me, seemingly working correctly in mega mode. See if you get anything after a few more minutes?

**Comment by TechnicJelle** [comment]
It restarts itself automatically, but the second time it runs, it fails to find the GPU again...

**Comment by Ly-Zxzy** [comment]
Ah yes, odd. Looking further, on startup I get the OOM errors only and not the cuDNN ones, so I assume the _Disabling cuDNN frontend for the following convolution:_ is what you'll want to look at solving to prevent the restart.

**Comment by fschwiet** [comment]
Yeah, my latest attempt failed. I'm going to try again later following all of @Ly-Zxzy instructions, in particular I didn't try 11.3 this time nor worry about the environment path.

@Ly-Zxzy can you share the output form lspci? It should include your video driver. The doc you linked in another section recommends using that to check your driver (section 2). I suspect that applies to people running Linux directly not via WSL. Per other docs at https://docs.nvidia.com/cuda/wsl-user-guide/index.html#getting-started-with-cuda-on-wsl it indicates a stub driver is installed that works with the native windows driver. Anyhow when I run it I get:

```
0482:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
7341:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
882a:00:00.0 3D controller: Microsoft Corporation Device 008e
cb34:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
```

I'm guessing 882a:00:00.0 is the stub driver the later documents talk about.

**Comment by raylin01** [comment]
For the people who have gotten it working on gpu, I was wondering how much CPU usage is being used. For some reason, my model (jax, pytorch all on gpu) is running on my gpu, but my cpu usage has 1 thread at 100%. Not sure if that is supposed to happen, but it seems to be the only thing that is bottlenecking the run. GPU usage is high, but not bottlenecked.

**Comment by rmartin16** [comment]
@raylin01, 3060 is pegged at 100% while a single thread on a gen4 i7 is also at 100%.
